### PR TITLE
Muon - Widget; Muon Grouping Table

### PR DESCRIPTION
**Pull request by TWJubb**
Refs #23642 

Requires all Muon core code to be merged.

This PR implements the grouping table widget, for use in the grouping tab of the new interface.

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        self._view.show()
+
+    def on_data_changed(self, notifier):
+        self._dataChangedNotifier = notifier
+
+    def notify_data_changed(self):
+        self._dataChangedNotifier()
+
+    def _is_edited_name_duplicated(self, new_name):
+        is_name_column_being_edited = self._view.grouping_table.currentColumn() == 0
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py]: sorry I do not understand this.

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    def disable_editing(self):
+        self._view.disable_editing()
+
+    def enable_editing(self):
+        self._view.enable_editing()
+
+    def add_group(self, group):
+        """Adds a group to the model and view"""
+        if self._view.num_rows() > 19:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py]: can we make this 19 and the 20 below use some sort of constant that we define at the top of the file?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def add_entry_to_table(self, row_entries):
+        assert len(row_entries) == self.grouping_table.columnCount()
+
+        row_position = self.grouping_table.rowCount()
+        self.grouping_table.insertRow(row_position)
+        for i, entry in enumerate(row_entries):
+            item = QtGui.QTableWidgetItem(entry)
+            if i == 0:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: can we make this indices constants at the top?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        self.remove_group_button.clicked.connect(slot)
+
+    def on_table_data_changed(self, slot):
+        self._on_table_data_changed = slot
+
+    def add_pair_requested(self):
+        selected_names = self.get_selected_group_names()
+        self.addPairRequested.emit(selected_names[0], selected_names[1])
+
+    def on_item_changed(self):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: what is the purpose of this?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    #
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def get_table_item_text(self, row, col):
+        return self.grouping_table.item(row, col).text()
+
+    def get_table_contents(self):
+        if self._updating:
+            return []
+        ret = [[None for _ in range(self.num_cols())]
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: what does the `_` mean?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+        for row in range(self.num_rows()):
+            for col in range(self.num_cols()):
+                item = self.grouping_table.item(row, col)
+                if col != 2:
+                    item.setFlags(QtCore.Qt.ItemIsSelectable |
+                                  QtCore.Qt.ItemIsEditable |
+                                  QtCore.Qt.ItemIsEnabled)
+                else:
+                    # number of detectors should remain un-editable
+                    item.setFlags(QtCore.Qt.ItemIsSelectable)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: all of the for and if statements in the view makes me feel uneasy.

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    # TESTS : Adding and removing groups
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def test_that_add_group_button_adds_group(self):
+        self.presenter.handle_add_group_button_clicked()
+        self.assertEqual(self.view.num_rows(), 1)
+        self.assertEqual(len(self.model.groups), 1)
+
+    def test_that_remove_group_button_removes_group(self):
+        self.add_three_groups_to_table()
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: check we have 3 too

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        self.view._get_selected_row_indices = mock.Mock(return_value=[0, 2])
+
+        self.presenter.handle_remove_group_button_clicked()
+
+        self.assertEqual(self.view.get_table_item_text(0, 0), "my_group_1")
+        self.assertEqual(self.model.group_names, ["my_group_1"])
+
+    def test_that_cannot_add_more_than_20_rows(self):
+        for i in range(21):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: can we make these constants at the top of the file?

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+    def test_that_trying_to_add_a_20th_row_gives_warning_message(self):
+        for i in range(21):
+            self.presenter.handle_add_group_button_clicked()
+
+        self.assertEqual(self.view.warning_popup.call_count, 1)
+
+    def test_that_remove_group_when_table_is_empty_does_not_throw(self):
+
+        self.presenter.handle_remove_group_button_clicked()
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: should this have an assert throws nothing and a check that the warning_popup call count is 0?

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    def test_context_menu_add_grouping_with_rows_selected_does_not_add_group(self):
+        self.presenter.handle_add_group_button_clicked()
+        self.view._get_selected_row_indices = mock.Mock(return_value=[0])
+
+        self.view.contextMenuEvent(0)
+
+        self.assertFalse(self.view.add_group_action.isEnabled())
+
+    def test_context_menu_remove_grouping_with_no_rows_selected_removes_last_row(self):
+        for i in range(3):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: can we not use the function here to add 3 groups?

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        self.view.contextMenuEvent(0)
+        self.view.remove_group_action.triggered.emit(True)
+
+        self.assertEqual(len(self.model.groups), 2)
+        self.assertEqual(self.view.num_rows(), 2)
+        self.assertEqual(self.view.get_table_item_text(0, 0), "group_1")
+        self.assertEqual(self.view.get_table_item_text(1, 0), "group_2")
+
+    def test_context_menu_remove_grouping_removes_selected_rows(self):
+        for i in range(3):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: same as above

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        self.assertIn("new_name", self.model.group_names)
+
+    def test_that_if_invalid_name_given_warning_message_is_shown(self):
+        self.add_three_groups_to_table()
+        print(self.view.get_table_contents())
+        invalid_names = ["", "@", "name!", "+-"]
+
+        for invalid_name in invalid_names:
+            self.view.grouping_table.item(0, 0).setText(invalid_name)
+            print(self.view.get_table_contents())
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: we shouldnt have a print statement in a production code test

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+        self.view.grouping_table.setCurrentCell(0, 1)
+        self.view.grouping_table.item(0, 1).setText("20-25,10,5,4,3,2,1")
+
+        self.assertEqual(self.view.get_table_item_text(0, 1), "1-5,10,20-25")
+        self.assertEqual(self.model._data._groups["group_1"].detectors, [1, 2, 3, 4, 5, 10, 20, 21, 22, 23, 24, 25])
+
+    def test_that_if_detector_list_changed_that_number_of_detectors_updates(self):
+        self.presenter.handle_add_group_button_clicked()
+
```
Comment (AnthonyLim23) [inline scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: can we check the original number of detectors too?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py (hunk 1 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py]: Amended to hopefully make clearer

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py (hunk 4 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: Currently this does nothing. We may in the future want to take some action when an item is changed but in the short term probably easier to remove and then add back in if needed

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py (hunk 5 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: I think this means that you don't care about the output, this line is being used to create a list of lists of set size pre-filled with None. I have updated the code to hopefully not need this.

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py (hunk 7 above)
Comment (Mantid-Matthew) [reply scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: I have added a separate test to check that the helper function `add_three_groups_to_table` works correctly as it is used in multiple places.

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py (hunk 6 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: either that we do a few tests in here

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py (hunk 9 above)
Comment (Mantid-Matthew) [reply scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: I think the test will automatically fail if an exception is thrown. I have added the check that the message box does not display.

File: scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py (hunk 10 above)
Comment (Mantid-Matthew) [reply scripts/test/Muon/grouping_tab/grouping_table_presenter_test.py]: yes we can

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_model.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    @property
+    def groups(self):
+        return self._data.groups
+
+    @property
+    def group_names(self):
+        return list(self._data.group_names)
+
+    @property
+    def group_and_pair_names(self):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_model.py]: this seems out of place here as everything else only handles groups

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def add_entry_to_table(self, row_entries):
+        assert len(row_entries) == self.grouping_table.columnCount()
+
+        row_position = self.grouping_table.rowCount()
+        self.grouping_table.insertRow(row_position)
+        for i, entry in enumerate(row_entries):
+            item = QtGui.QTableWidgetItem(entry)
+            if group_table_columns[i] == 'group_name':
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: should use the `group_table_columns` at the top of the file

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py (hunk 6 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: can we add a test for this?

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_model.py (hunk 14 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_model.py]: The grouping table needs to know the pair names to ensure that it does not allow a group to be added with the same name as a pair.

File: scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py (hunk 15 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_view.py]: Yes

**Review by AnthonyLim23 — APPROVED** [review]

**Pull request merged**